universal-application-tool-0.0.1/app/services/program/predicate/ConcretePredicateExpressionNode.java

```diff
@@ -2,11 +2,16 @@
 
 import com.fasterxml.jackson.annotation.JsonSubTypes;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.google.common.collect.ImmutableList;
+import services.question.types.QuestionDefinition;
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
 @JsonSubTypes({@JsonSubTypes.Type(value = LeafOperationExpressionNode.class, name = "leaf")})
 public interface ConcretePredicateExpressionNode {
 
   /** Returns the type of this node, as a {@link PredicateExpressionNodeType}. */
   PredicateExpressionNodeType getType();
+
+  /** Returns a human-readable interpretation of this node. */
+  String toDisplayString(ImmutableList<QuestionDefinition> questions);
```

Comment:
suggestion (nit and non-blocking): Consider briefly explaining why it's necessary to pass in a list of questions here? It's a little unintuitive why it needs to be a list rather than just a single question.

Reply:
Great point! I added documentation on this

Reply:
Looks so good!